### DISCUSSION TITLE
Get epoch rewards for a block from node.

### DISCUSSION THREAD

**Discussion by shaswatsaloni** [celo-org/celo-blockchain#2256] [Help (Q&A)]
Hi Team,

I have a rosetta as well as archival Celo Node. There i am getting the normal transactions with the query:

{
"method": "eth_getBlockByNumber",
"params": [
"0x556E00", //blocknumber
true
],
"id": 1,
"jsonrpc": "2.0"
}

But here the epoch transactions are not present, by which i want to have the voting rewards. 

But if i want to query the rosetta or archival node for this epoch transactions/rewards, how would i get that?

@eelanagaraj  could you please help us here.

Thanks,
Saloni

**Comment by carterqw2** [comment]
Hey @shaswatsaloni, could you please check out this [repo](https://github.com/celo-org/epochs) for an explainer on how epoch transactions work and how to fetch them.

**Comment by shaswatsaloni** [comment]
Hey @carterqw2 , that repo was really helpful, thanks.

One more query here, i want to get the epoch rewards for the address : 0x9efecf85d7e668c0c3c7a778a1164b064234f96d

So, here i can see it has epoch transactions in block: [5598720]

the explorer link : https://explorer.celo.org/mainnet/address/0x9efeCf85d7E668c0C3c7a778A1164b064234f96d/epoch-transactions

![image](https://github.com/celo-org/celo-blockchain/assets/132393582/7c1f2203-019d-4d3b-ba26-d56dbab33f8f)

When i am calling the POST method for this particular block:

{
    "method": "eth_getLogs",
    "params": [
        {
            "fromBlock": "0x556E00",
            "toBlock": "0x556E00"
        }
    ],
    "id": 1,
    "jsonrpc": "2.0"
}

I am not able to find the particular address (0x9efecf85d7e668c0c3c7a778a1164b064234f96d) in the response JSON.

how would i calculate and get the epoch rewards here?

Thanks, 
Saloni

**Comment by carterqw2** [comment]
Hi @shaswatsaloni! You're trying to get individual voter rewards, please refer to [this section](https://github.com/celo-org/epochs?tab=readme-ov-file#individual-voter-rewards) of the explainer for more details. Celo Explorer has to perform additional calculations to be able to display individual voter rewards that are not directly available through logs. I hope that helps.

**Comment by shaswatsaloni** [comment]
Okay, understood.

So, there is no simple way that we an fetch epoch transactions/ voting rewards for a celo address.

And, is there any way we can get the epoch transactions for an address instead of a block.

Thanks, 
Saloni

**Comment by carterqw2** [comment]
We provide an [RPC API](https://explorer.celo.org/mainnet/api-docs) in Celo Explorer, if that works for you (search for `getvoterrewards`).

https://explorer.celo.org/mainnet/api?module=epoch&action=getvoterrewards&voterAddress=0x9efecf85d7e668c0c3c7a778a1164b064234f96d

```
curl -X GET "https://explorer.celo.org/mainnet/api?module=epoch&action=getvoterrewards&voterAddress=0x9efecf85d7e668c0c3c7a778a1164b064234f96d" -H "accept: application/json"
```

**Comment by shaswatsaloni** [comment]
Yes, this i have. I just wanted to query my node if its at all possible?

**Comment by carterqw2** [comment]
Yeah, unfortunately, there is no easy way, you'd have to fetch epoch blocks and perform calculations [described](https://github.com/celo-org/epochs?tab=readme-ov-file#individual-voter-rewards) in the explainer manually.

**Comment by shaswatsaloni** [comment]
Hi @carterqw2 , from this RPCAPI https://explorer.celo.org/mainnet/api?module=epoch&action=getvoterrewards&voterAddress=0x9efecf85d7e668c0c3c7a778a1164b064234f96d, we are getting the rewards based on the address.
Is there a way, we can get it for a block? And, is there any limitation for the number of requests we make for this API?

Thanks,
Saloni.

**Reply by carterqw2** [reply]
Do I understand you correctly that you want to fetch all voter rewards split by account for a given epoch block? If so, then we don't have such an API.

In theory, you could implement something like:
1. Get all the [`ValidatorGroupVoteActivated`](https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/governance/Election.sol#L128) events to get all the unique voters accounts.
2. For all the voter addresses, get their rewards using the RPC API endpoint above.

It would be a bit chatty but would work. I'm not sure what the current rate limit values are but please avoid abusing the system.
